Organize application code into modules

dylanmccall opened the pull request:
https://phabricator.endlessm.com/T34007

manuq committed:
Version: Add Explore plugin version to app version

And also shorten the Kolibri version if it comes from a git branch (ie
it's way too long).

https://phabricator.endlessm.com/T34102

dylanmccall committed:
Add Activity and Service classes to manage state

https://phabricator.endlessm.com/T34007

dylanmccall commented:
This needs some rebasing, but it is now in a working state! There should be no functional changes from the `develop` branch here.

manuq commented on src/kolibri_android/android_utils.py:
```diff
@@ -22,7 +22,8 @@
 from jnius import JavaException
 from jnius import jnius
-from runnable import Runnable
+
+from .runnable import Runnable
```
I guess this could be explained as "Use relative imports for inner modules" and all similar changes be split into a separate commit.

manuq commented on src/kolibri_android/kolibri_utils.py:
```diff
@@ -0,0 +1,139 @@
+
+    # Don't set this if the retrieved id is falsy, too short, or a specific
+    # id that is known to be hardcoded in many devices.
+    if node_id and len(node_id) >= 16 and node_id != "9774d56d682e549c":
+        os.environ["MORANGO_NODE_ID"] = node_id
```
Can you explain what is this about? Ideally it should be in a separate commit as this is new stuff not a refactor. What is affected by setting the `MORANGO_NODE_ID` environment variable? I don't see it in Kolibri's option.ini or anywhere in its source code.

dylanmccall replied:
This comes from [initialization.py](https://github.com/endlessm/kolibri-installer-android/blob/develop/src/initialization.py#L77-L80). It was added in learningequality/kolibri-installer-android#88, and it causes Kolibri's database replication system to reuse the Android device ID instead of its default behaviour. I will add another commit that is just moving stuff out of initialize.py so that the provenance of stuff like this is clearer in the commit history.

manuq commented on src/kolibri_android/kolibri_utils.py:
```diff
@@ -0,0 +1,139 @@
+]
+
+
+def init_kolibri(content_fallback_dirs=None, **kwargs):
```
Nice to see this moved to the utils module.

manuq commented on src/kolibri_android/application.py:
```diff
@@ -0,0 +1,42 @@
+from android.activity import register_activity_lifecycle_callbacks
+
+
+class BaseActivity(object):
```
Nice. We don't have 2 classes inheriting it but the abstraction looks ready for upstreaming or adding a second activity.

manuq commented on src/kolibri_android/main_activity/__main__.py:
```diff
@@ -0,0 +1,14 @@
+initialize()
+
+
+def main():
```
Nice and clean main. :+1:

manuq commented on src/kolibri_android/main_activity/activity.py:
```diff
@@ -0,0 +1,185 @@
+    def run(self):
+        self.load_url("file:///android_asset/_load.html")
+
+        while True:
```
I wish we'll find a better way for the main loop. But it's good keep it as-is for now.

dylanmccall replied:
Yeah, I agree. I think we can probably set up a message queue or something now that there are some more bones to work with. I figure that overlaps with the other problem where we need to wait for external storage to be available before doing any of it.

manuq commented on src/kolibri_android/remoteshell_service/service.py:
```diff
@@ -0,0 +1,125 @@
+from ..application import BaseService
+
+
+class RemoteShellService(BaseService):
```
Do you know if this remote shell works at all? I never tried it but it would be good for debugging.

dylanmccall replied:
It didn't originally even have a main function, which is _weird_. It has one now, but the service isn't being run anywhere either. I thought about doing that, but it felt out of scope :) It would definitely be great to get this up and running for real!

manuq reviewed (changes requested):
Exciting changes!

One request: This comes into a single big commit that lacks a description. Of course with refactors is not always easy to split the changes. I left some ideas. Could you also justify them in the commit description? In particular would this improve the app lifecycle in any ways as Simon was expecting? Like waiting for the storage to be ready? Or is a pure refactor and we should expect the exact same behavior as before? Also please note this in the ticket as "acceptance criteria" for verifying.

manuq commented on src/kolibri_android/kolibri_utils.py:
```diff
@@ -0,0 +1,139 @@
+    _monkeypatch_whitenoise()
+    _init_kolibri_env(content_fallback_dirs)
+
+    from kolibri.utils.main import initialize
```
One more question. I see a lot of imports inside functions in the `kolibri_utils.py` file. However the Python PEP8 recommends to put all imports at the top of the file.

dylanmccall replied:
Yeah, these are to deal with Kolibri imports that require various environment setup to happen first. In this case, importing `initialize` also means importing `conf`, which does a lot of things on import.

Probably worth going through them again and making sure it only happens where it's absolutely necessary. Alternatively, could do lazy imports to make things nicer.

But that bit in particular I'll fix right now. Imports in the middle of a function are _extra_ bad :)

manuq replied:
Usually these kind of lazy importing are an indicator that the module needs to be split further. What do you think about doing that?

dylanmccall replied:
Yeah, I'd like to get things to a state where we're working within Kolibri's world such that it doesn't _matter_ if importing has side-effects. We should be configuring our environment by setting things in `kolibri.conf` - which we can do any time before running `initialize()` - not by using environment variables.

dylanmccall committed:
Move application code to a Python package

https://phabricator.endlessm.com/T34007

dylanmccall committed:
Add a function to set up the Python environment

Instead of setting up the environment when initialize.py is imported, we
move that code inside a function that is explicitly called by the main
function.

https://phabricator.endlessm.com/T34007

dylanmccall committed:
Add get_android_node_id to android_utils

https://phabricator.endlessm.com/T34007

dylanmccall committed:
Move Kolibri environment setup to another module

https://phabricator.endlessm.com/T34007

dylanmccall commented:
Okay, I went through and split this work into some more concrete pieces for the commit log - and fixed a few small cosmetic issues and things that I didn't mean to change here. Much easier to spot with the commits better organized :) The last commit is still pretty gigantic, seeing as it involves indenting a bunch of code slightly differently and moving functions to different files, but I think it's nicer to look at now.

manuq commented on src/kolibri_android/initialization.py:
```diff
@@ -71,9 +69,7 @@ def initialize():
-    Secure = autoclass("android.provider.Settings$Secure")
-
-    node_id = Secure.getString(get_activity().getContentResolver(), Secure.ANDROID_ID)
+    node_id = get_android_node_id()
```
:+1:
